File: clojure/99_problems_in_lisp/ninety_nine_problems.clj
;; https://chatgpt.com/c/6884d93b-6d6c-832e-b84c-163fd3fa4f58
;; -*- clojure -*-
;; Solutions to selected Ninety-Nine Lisp problems in Clojure.
;;
;; This namespace implements a subset of the classic "Ninety-Nine Lisp Problems"
;; using idiomatic Clojure.  The functions cover problems related to
;; list processing, combinatorics, and basic number theory.  Each problem
;; is accompanied by a unit test in `ninety_nine_problems_test.clj`.  See the
;; original problem descriptions at the University of Campinas site for
;; details on each specification【970221147506275†L6-L20】【970221147506275†L29-L46】.

(ns ninety-nine-problems.core
  (:require [clojure.set :as set]))

;; Problem P01: Find the last element of a list【970221147506275†L6-L9】
(defn my-last [coll]
  (if (empty? coll)
    nil
    (let [rest (rest coll)]
      (if (empty? rest)
        (first coll)
        (recur rest)))))

;; P02: Find the last but one element of a list, return the last two elements【970221147506275†L11-L14】
(defn my-but-last [coll]
  (if (or (empty? coll) (< (count coll) 2))
    nil
    [(nth coll (- (count coll) 2)) (last coll)]))

;; P03: Find the K'th element of a list (1-indexed)【970221147506275†L16-L20】
(defn element-at [coll k]
  (nth coll (dec k)))

;; P04: Count the number of elements of a list【970221147506275†L22-L23】
(defn my-count [coll]
  (reduce (fn [acc _] (inc acc)) 0 coll))

;; P05: Reverse a list【970221147506275†L24-L25】
(defn my-reverse [coll]
  (reduce (fn [acc x] (cons x acc)) '() coll))

;; P06: Find out whether a list is a palindrome【970221147506275†L26-L28】
(defn palindrome? [coll]
  (= (seq coll) (reverse coll)))

;; P07: Flatten a nested list structure【970221147506275†L29-L36】
(defn flatten-list [coll]
  (filter (complement sequential?)
          (tree-seq sequential? seq coll)))

;; P08: Eliminate consecutive duplicates of list elements【970221147506275†L39-L46】
(defn compress [coll]
  (when (seq coll)
    (let [x (first coll)]
      (cons x (lazy-seq (compress (drop-while #(= % x) (rest coll))))))))

;; P09: Pack consecutive duplicates of list elements into sublists【970221147506275†L47-L53】
(defn pack [coll]
  (when (seq coll)
    (let [x (first coll)
          same (take-while #(= % x) coll)
          rest (drop-while #(= % x) coll)]
      (cons same (pack rest)))))

;; P10: Run-length encoding of a list【970221147506275†L55-L62】
(defn encode [coll]
  (map #(vector (count %) (first %)) (pack coll)))

;; P11: Modified run-length encoding【970221147506275†L64-L71】
(defn encode-modified [coll]
  (map (fn [[n x]] (if (= n 1) x [n x])) (encode coll)))

;; P12: Decode a run-length encoded list【970221147506275†L73-L75】
(defn decode [coll]
  (apply concat
         (map (fn [e] (if (sequential? e)
                        (repeat (first e) (second e))
                        [e]))
              coll)))

;; P13: Run-length encoding of a list (direct solution)【970221147506275†L77-L85】
(defn encode-direct [coll]
  (when (seq coll)
    (let [x (first coll)
          same (take-while #(= % x) coll)
          rest (drop-while #(= % x) coll)
          head (if (= 1 (count same)) x [(count same) x])]
      (cons head (encode-direct rest)))))

;; P14: Duplicate the elements of a list【970221147506275†L87-L90】
(defn dupli [coll]
  (mapcat #(repeat 2 %) coll))

;; P15: Replicate the elements of a list a given number of times【970221147506275†L92-L95】
(defn repli [coll n]
  (mapcat #(repeat n %) coll))

;; P16: Drop every N'th element from a list【970221147506275†L97-L101】
(defn drop-every [coll n]
  (->> coll
       (map-indexed vector)
       (remove (fn [[idx _]] (zero? (mod (inc idx) n))))
       (map second)))

;; P17: Split a list into two parts【970221147506275†L102-L107】
(defn split [coll n]
  [(take n coll) (drop n coll)])

;; P18: Extract a slice from a list (inclusive)【970221147506275†L109-L116】
(defn slice [coll i k]
  (->> coll (drop (dec i)) (take (- (inc k) i))))

;; P19: Rotate a list N places to the left (negative rotates right)【970221147506275†L118-L125】
(defn rotate [coll n]
  (let [len (count coll)
        n (if (zero? len) 0 (mod n len))]
    (concat (drop n coll) (take n coll))))

;; P20: Remove the K'th element from a list【970221147506275†L129-L133】
(defn remove-at [coll k]
  (concat (take (dec k) coll) (drop k coll)))

;; P21: Insert an element at a given position into a list【970221147506275†L134-L137】
(defn insert-at [elem coll k]
  (concat (take (dec k) coll) [elem] (drop (dec k) coll)))

;; P22: Create a list containing all integers within a given range【970221147506275†L139-L145】
(defn range-list [start end]
  (if (<= start end)
    (range start (inc end))
    (range start (dec end) -1)))

;; P23: Extract a given number of randomly selected elements from a list【970221147506275†L147-L153】
(defn rnd-select [coll n]
  (take n (shuffle coll)))

;; P24: Lotto: Draw N different random numbers from the set 1..M【970221147506275†L155-L161】
(defn lotto-select [n m]
  (rnd-select (range-list 1 m) n))

;; P25: Generate a random permutation of the elements of a list【970221147506275†L163-L169】
(defn rnd-permu [coll]
  (shuffle coll))

;; P26: Generate the combinations of K distinct objects chosen from the N elements of a list【970221147506275†L170-L179】
(defn combinations [k coll]
  (cond
    (= k 0) '(())
    (empty? coll) '()
    :else (let [x (first coll)
                xs (rest coll)
                with-x (map #(cons x %) (combinations (dec k) xs))
                without-x (combinations k xs)]
            (concat with-x without-x))))

;; P27: Group the elements of a set into disjoint subsets【970221147506275†L181-L206】
(defn group3 [coll]
  (for [a (combinations 2 coll)
        b (combinations 3 (set/difference (set coll) (set a)))
        :let [remaining (set/difference (set coll) (set (concat a b)))]
        c [(vec remaining)]]
    [a b c]))

(defn group [coll sizes]
  (if (empty? sizes)
    '(())
    (let [n (first sizes)]
      (for [comb (combinations n coll)
            rest (group (vec (set/difference (set coll) (set comb))) (rest sizes))]
        (cons comb rest)))))

;; P28: Sorting a list of lists according to length of sublists【970221147506275†L206-L219】
(defn lsort [coll]
  (sort-by count coll))

(defn lfsort [coll]
  (let [freq (frequencies (map count coll))]
    (sort-by #(freq (count %)) coll)))

;; Arithmetic and number theory problems

;; P31: Determine whether a given integer number is prime【970221147506275†L230-L235】
(defn is-prime? [n]
  (and (> n 1)
       (not-any? #(zero? (mod n %)) (range 2 (inc (Math/sqrt n))))))

;; P32: Determine the greatest common divisor of two positive integer numbers【970221147506275†L237-L242】
(defn gcd [a b]
  (if (zero? b)
    a
    (recur b (mod a b))))

;; P33: Determine whether two positive integer numbers are coprime【970221147506275†L244-L248】
(defn coprime? [a b]
  (= 1 (gcd a b)))

;; P34: Calculate Euler's totient function phi(m) (naive)【970221147506275†L250-L258】
(defn totient-phi [m]
  (count (filter #(coprime? % m) (range 1 m))))

;; P35: Determine the prime factors of a given positive integer【970221147506275†L266-L270】
(defn prime-factors [n]
  (letfn [(factor [n f acc]
            (cond
              (= n 1) acc
              (zero? (mod n f)) (recur (/ n f) f (conj acc f))
              :else (recur n (inc f) acc)))]
    (factor n 2 [])))

;; P36: Determine the prime factors of a given positive integer (with multiplicity)【970221147506275†L272-L276】
(defn prime-factors-mult [n]
  (->> (prime-factors n)
       (frequencies)
       (map (fn [[p m]] [p m]))
       (sort-by first)))

;; P37: Calculate Euler's totient function phi(m) (improved)【970221147506275†L280-L283】
(defn phi [m]
  (reduce (fn [acc [p m]]
            (* acc (- (Math/pow p m) (Math/pow p (dec m)))))
          1
          (prime-factors-mult m)))

;; Note: Additional problems (P38–P99) are not implemented here due to time constraints.

File: clojure/99_problems_in_lisp/ninety_nine_problems.py
# https://chatgpt.com/c/6884d93b-6d6c-832e-b84c-163fd3fa4f58
"""
Python implementations for a subset of the classic Ninety‑Nine Lisp Problems.

This module defines functions corresponding to many of the problems from the
University of Campinas list【970221147506275†L6-L20】【970221147506275†L26-L46】.  The focus here is on
list processing (P01–P28), arithmetic (P31–P37) and logic/codes (P46–P50).

Each problem is implemented in a straightforward, Pythonic way.  See the
accompanying `ninety_nine_problems_test.py` for unit tests verifying
correctness.  Problems beyond P50 are not implemented in this module.
"""

from __future__ import annotations

import math
import random
import itertools
from collections import Counter
from typing import Any, Iterable, List, Sequence, Tuple, Dict


# -----------------------------------------------------------------------------
# List problems

def my_last(seq: Sequence[Any]) -> Any | None:
    """Return the last element of a sequence or None if empty."""
    return seq[-1] if seq else None


def my_but_last(seq: Sequence[Any]) -> List[Any] | None:
    """Return a list containing the last two elements of the sequence."""
    if len(seq) < 2:
        return None
    return [seq[-2], seq[-1]]


def element_at(seq: Sequence[Any], k: int) -> Any:
    """Return the Kth element of seq (1-indexed)."""
    return seq[k - 1]


def my_count(seq: Sequence[Any]) -> int:
    """Return the number of elements in seq."""
    return sum(1 for _ in seq)


def my_reverse(seq: Sequence[Any]) -> List[Any]:
    """Return a new list with the elements of seq in reverse order."""
    return list(reversed(seq))


def palindrome(seq: Sequence[Any]) -> bool:
    """Return True if seq is a palindrome."""
    s = list(seq)
    return s == s[::-1]


def flatten_list(seq: Iterable[Any]) -> List[Any]:
    """Flatten a nested list structure into a single flat list."""
    result: List[Any] = []
    for x in seq:
        if isinstance(x, (list, tuple)):
            result.extend(flatten_list(x))
        else:
            result.append(x)
    return result


def compress(seq: Sequence[Any]) -> List[Any]:
    """Eliminate consecutive duplicates from seq."""
    result: List[Any] = []
    prev = object()
    for x in seq:
        if x != prev:
            result.append(x)
            prev = x
    return result


def pack(seq: Sequence[Any]) -> List[List[Any]]:
    """Pack consecutive duplicates into sublists."""
    if not seq:
        return []
    result: List[List[Any]] = []
    current: List[Any] = [seq[0]]
    for x in seq[1:]:
        if x == current[-1]:
            current.append(x)
        else:
            result.append(current)
            current = [x]
    result.append(current)
    return result


def encode(seq: Sequence[Any]) -> List[Tuple[int, Any]]:
    """Run-length encode seq using pack."""
    return [(len(group), group[0]) for group in pack(seq)]


def encode_modified(seq: Sequence[Any]) -> List[Any]:
    """Modified run-length encoding: singletons are copied as is."""
    result: List[Any] = []
    for count, item in encode(seq):
        if count == 1:
            result.append(item)
        else:
            result.append((count, item))
    return result


def decode(seq: Sequence[Any]) -> List[Any]:
    """Decode a run-length encoded sequence."""
    result: List[Any] = []
    for elem in seq:
        if isinstance(elem, tuple):
            count, item = elem
            result.extend([item] * count)
        else:
            result.append(elem)
    return result


def encode_direct(seq: Sequence[Any]) -> List[Any]:
    """Direct run-length encoding without creating intermediate lists."""
    if not seq:
        return []
    result: List[Any] = []
    count = 1
    prev = seq[0]
    for x in seq[1:]:
        if x == prev:
            count += 1
        else:
            result.append((count, prev) if count > 1 else prev)
            prev = x
            count = 1
    result.append((count, prev) if count > 1 else prev)
    return result


def dupli(seq: Sequence[Any]) -> List[Any]:
    """Duplicate each element of seq."""
    result: List[Any] = []
    for x in seq:
        result.extend([x, x])
    return result


def repli(seq: Sequence[Any], n: int) -> List[Any]:
    """Replicate each element of seq n times."""
    result: List[Any] = []
    for x in seq:
        result.extend([x] * n)
    return result


def drop_every(seq: Sequence[Any], n: int) -> List[Any]:
    """Drop every n-th element from seq."""
    result: List[Any] = []
    for idx, x in enumerate(seq, start=1):
        if idx % n != 0:
            result.append(x)
    return result


def split(seq: Sequence[Any], n: int) -> Tuple[List[Any], List[Any]]:
    """Split seq into two parts; the first part has n elements."""
    return (list(seq[:n]), list(seq[n:]))


def slice_(seq: Sequence[Any], i: int, k: int) -> List[Any]:
    """Return a slice from the ith to the kth element (inclusive, 1-indexed)."""
    return list(seq[i - 1 : k])


def rotate(seq: Sequence[Any], n: int) -> List[Any]:
    """Rotate seq n places to the left (negative rotates right)."""
    n = n % len(seq) if seq else 0
    return list(seq[n:] + seq[:n])


def remove_at(seq: Sequence[Any], k: int) -> List[Any]:
    """Remove the k-th element from seq (1-indexed)."""
    return list(seq[: k - 1] + seq[k:])


def insert_at(elem: Any, seq: Sequence[Any], k: int) -> List[Any]:
    """Insert elem at position k in seq (1-indexed)."""
    return list(seq[: k - 1] + [elem] + seq[k - 1 :])


def range_list(start: int, end: int) -> List[int]:
    """Create a list containing all integers from start to end inclusive."""
    step = 1 if start <= end else -1
    return list(range(start, end + step, step))


def rnd_select(seq: Sequence[Any], n: int) -> List[Any]:
    """Return a list of n randomly selected elements from seq."""
    return random.sample(list(seq), n)


def lotto_select(n: int, m: int) -> List[int]:
    """Return n different random numbers selected from 1..m."""
    return rnd_select(range_list(1, m), n)


def rnd_permu(seq: Sequence[Any]) -> List[Any]:
    """Return a random permutation of seq."""
    lst = list(seq)
    random.shuffle(lst)
    return lst


def combinations(k: int, seq: Sequence[Any]) -> List[List[Any]]:
    """Generate all combinations of k distinct elements chosen from seq."""
    return [list(comb) for comb in itertools.combinations(seq, k)]


def group3(seq: Sequence[Any]) -> List[List[List[Any]]]:
    """Group elements of seq into subsets of sizes 2, 3 and the remainder."""
    result: List[List[List[Any]]] = []
    seq_set = list(seq)
    for a in combinations(2, seq_set):
        rem1 = [x for x in seq_set if x not in a]
        for b in combinations(3, rem1):
            rem2 = [x for x in rem1 if x not in b]
            result.append([a, b, rem2])
    return result


def group(seq: Sequence[Any], sizes: Sequence[int]) -> List[List[List[Any]]]:
    """Group the elements of ``seq`` into disjoint subsets of the specified sizes.

    The order of the resulting groups is irrelevant; duplicate partitions (where
    the same subsets appear in a different order) are removed.  Each element of
    ``seq`` appears in exactly one subset.  For example, grouping four items
    into two pairs yields three unique groupings rather than six permutations.

    This implementation generates partitions recursively using combinations and
    then deduplicates the results by sorting the groups and using a set to
    eliminate equivalent partitions.
    """
    if not sizes:
        return [[]]
    first, *rest = sizes
    partial_results: List[List[List[Any]]] = []
    for comb in combinations(first, seq):
        remaining = [x for x in seq if x not in comb]
        for rest_groups in group(remaining, rest):
            partial_results.append([list(comb)] + rest_groups)
    # Deduplicate partitions: sort each subgroup and then sort the list of groups.
    unique: Dict[Tuple[Tuple[Any, ...], ...], List[List[Any]]] = {}
    for groups in partial_results:
        # Normalize each subgroup and the order of subgroups for hashing.
        normalized = tuple(sorted(tuple(sorted(subgrp)) for subgrp in groups))
        if normalized not in unique:
            unique[normalized] = groups
    return list(unique.values())


def lsort(seq: Sequence[Sequence[Any]]) -> List[List[Any]]:
    """Sort a list of lists according to length of sublists."""
    return sorted([list(s) for s in seq], key=len)


def lfsort(seq: Sequence[Sequence[Any]]) -> List[List[Any]]:
    """Sort a list of lists according to the frequency of their lengths."""
    lengths = [len(s) for s in seq]
    freq = Counter(lengths)
    return sorted([list(s) for s in seq], key=lambda s: freq[len(s)])


# -----------------------------------------------------------------------------
# Arithmetic problems


def is_prime(n: int) -> bool:
    """Return True if n is prime."""
    if n <= 1:
        return False
    if n <= 3:
        return True
    if n % 2 == 0 or n % 3 == 0:
        return False
    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i + 2) == 0:
            return False
        i += 6
    return True


def gcd(a: int, b: int) -> int:
    """Compute the greatest common divisor using Euclid's algorithm."""
    while b:
        a, b = b, a % b
    return abs(a)


def coprime(a: int, b: int) -> bool:
    """Return True if a and b are coprime."""
    return gcd(a, b) == 1


def totient_phi(m: int) -> int:
    """Euler's totient function phi(m), naive implementation."""
    return sum(1 for r in range(1, m) if coprime(r, m))


def prime_factors(n: int) -> List[int]:
    """Return the prime factors of n in ascending order."""
    factors: List[int] = []
    f = 2
    while n > 1:
        while n % f == 0:
            factors.append(f)
            n //= f
        f += 1
    return factors


def prime_factors_mult(n: int) -> List[Tuple[int, int]]:
    """Return the prime factors and their multiplicities."""
    c = Counter(prime_factors(n))
    return sorted(list(c.items()))


def phi_improved(m: int) -> int:
    """Improved Euler's totient using prime factorisation."""
    result = 1
    for p, k in prime_factors_mult(m):
        result *= (p ** k - p ** (k - 1))
    return result


# -----------------------------------------------------------------------------
# Logic and Codes


def nand(a: bool, b: bool) -> bool:
    return not (a and b)


def nor(a: bool, b: bool) -> bool:
    return not (a or b)


def xor(a: bool, b: bool) -> bool:
    return a != b


def impl(a: bool, b: bool) -> bool:
    return (not a) or b


def equ(a: bool, b: bool) -> bool:
    return a == b


# Helper functions for evaluating prefix boolean expressions used in truth tables.
def _parse_prefix(tokens: List[str]) -> Any:
    """Parse a list of tokens representing a prefix expression with parentheses.

    Returns a nested list structure representing the abstract syntax tree (AST).
    Each sublist begins with an operator followed by its operands.  Atoms
    (variables or constants) are returned as strings.  This parser handles
    parentheses and arbitrary nesting; it assumes that the input is well‑formed.
    """
    if not tokens:
        raise ValueError("Unexpected end of tokens during parsing")
    token = tokens.pop(0)
    if token == '(':  # start of an expression
        if not tokens:
            raise ValueError("Missing operator after '('")
        op = tokens.pop(0)
        args: List[Any] = []
        while tokens and tokens[0] != ')':
            args.append(_parse_prefix(tokens))
        if not tokens or tokens[0] != ')':
            raise ValueError("Missing closing ')' in expression")
        tokens.pop(0)  # consume ')'
        return [op] + args
    elif token == ')':
        raise ValueError("Unexpected ')' in expression")
    else:
        return token


def _eval_ast(ast: Any, env: Dict[str, Any]) -> Any:
    """Evaluate an AST produced by ``_parse_prefix`` in the given environment.

    The environment ``env`` maps variable names and operator symbols to
    corresponding boolean values or functions.  The evaluator supports unary
    operator ``not`` and n‑ary operators such as ``and``, ``or``, ``nand``,
    ``nor``, ``xor``, ``impl`` and ``equ``.  Atoms that match keys in ``env``
    are looked up directly; literal "True" and "False" are interpreted as
    boolean constants.
    """
    # If the AST is a list, evaluate recursively.
    if isinstance(ast, list):
        op = ast[0]
        # Evaluate all operand expressions first
        operands = [ _eval_ast(arg, env) for arg in ast[1:] ]
        # Fetch the operator function from the environment
        if op not in env:
            raise KeyError(f"Unknown operator: {op}")
        func = env[op]
        return func(*operands)
    # Otherwise, it's an atom (variable or constant)
    if isinstance(ast, str):
        if ast == 'True':
            return True
        if ast == 'False':
            return False
        if ast in env:
            return env[ast]
    # If nothing matches, raise
    raise KeyError(f"Unbound symbol: {ast}")


def table(var1: str, var2: str, expr) -> List[List[Any]]:
    """Generate a truth table for a two‑variable logical expression.

    ``expr`` may be either a callable taking two boolean arguments or a
    prefix expression string using parentheses and operator names as defined
    in the Ninety‑Nine problems, such as ``"(and (or A B) (nand A B))"``.  The
    function evaluates the expression for all four combinations of truth
    assignments to ``var1`` and ``var2`` and returns a list of rows of the
    form ``[value_of_var1, value_of_var2, result]``.
    """
    rows: List[List[Any]] = []
    # Determine how to evaluate the expression
    if callable(expr):
        def eval_expr(a: bool, b: bool) -> Any:
            return expr(a, b)
    elif isinstance(expr, str):
        # Parse the prefix expression into an AST once
        tokens = expr.replace('(', ' ( ').replace(')', ' ) ').split()
        ast = _parse_prefix(tokens)
        # Build an environment mapping operator names and variable names to functions/values
        def eval_expr(a: bool, b: bool) -> Any:
            env = {
                var1: a,
                var2: b,
                'nand': nand,
                'nor': nor,
                'xor': xor,
                'impl': impl,
                'equ': equ,
                'and': lambda x, y: x and y,
                'or': lambda x, y: x or y,
                'not': lambda x: not x,
            }
            return _eval_ast(ast, env)
    else:
        raise TypeError("expr must be either a callable or a string")
    # Evaluate for all combinations of True/False for var1 and var2
    for a in [True, False]:
        for b in [True, False]:
            rows.append([a, b, eval_expr(a, b)])
    return rows


def assignments(vars_: Sequence[str]) -> List[Dict[str, bool]]:
    """Generate all assignments of True/False for the given variable names."""
    if not vars_:
        return [{}]
    first, *rest = vars_
    sub = assignments(rest)
    result: List[Dict[str, bool]] = []
    for a in sub:
        result.append({**a, first: True})
        result.append({**a, first: False})
    return result


def table_n(vars_: Sequence[str], expr) -> List[List[Any]]:
    """Generate a truth table for a logical expression over multiple variables.

    ``vars_`` is a sequence of variable names (strings).  ``expr`` may be a
    callable accepting the same number of boolean arguments as there are
    variables, or a prefix expression string using parentheses and the
    operators defined in the Ninety‑Nine problems.  The function returns a
    list of rows, each consisting of the assigned boolean values for the
    variables followed by the result of the expression.
    """
    rows: List[List[Any]] = []
    if callable(expr):
        def eval_expr(*args: bool) -> Any:
            return expr(*args)
    elif isinstance(expr, str):
        tokens = expr.replace('(', ' ( ').replace(')', ' ) ').split()
        ast = _parse_prefix(tokens)
        def eval_expr(*args: bool) -> Any:
            env = {var: val for var, val in zip(vars_, args)}
            env.update({
                'nand': nand,
                'nor': nor,
                'xor': xor,
                'impl': impl,
                'equ': equ,
                'and': lambda x, y: x and y,
                'or': lambda x, y: x or y,
                'not': lambda x: not x,
            })
            return _eval_ast(ast, env)
    else:
        raise TypeError("expr must be either a callable or a string")
    for assign in assignments(list(vars_)):
        values = [assign[var] for var in vars_]
        rows.append(values + [eval_expr(*values)])
    return rows


def gray(n: int) -> List[str]:
    """Return the n-bit Gray code sequence."""
    if n < 0:
        raise ValueError("n must be non-negative")
    if n == 0:
        return [""]
    if n == 1:
        return ["0", "1"]
    prev = gray(n - 1)
    return ["0" + x for x in prev] + ["1" + x for x in reversed(prev)]


def huffman(freqs: List[Tuple[Any, int]]) -> List[Tuple[Any, str]]:
    """Return the Huffman code for a frequency table."""
    # Build initial forest of leaf nodes
    forest = [({'symbol': sym, 'freq': freq}, None, None) for sym, freq in freqs]
    # Each node is a tuple (node, left, right) where node is dict containing freq and symbol(s)
    while len(forest) > 1:
        # Sort forest by frequency
        forest.sort(key=lambda t: t[0]['freq'])
        (node1, left1, right1), (node2, left2, right2), *rest = forest
        merged = ({'symbol': (node1['symbol'], node2['symbol']),
                   'freq': node1['freq'] + node2['freq']},
                  (node1, left1, right1),
                  (node2, left2, right2))
        forest = rest + [merged]
    tree = forest[0]

    codes: Dict[Any, str] = {}
    def assign_codes(node, left, right, prefix: str):
        sym = node['symbol']
        if not isinstance(sym, tuple):
            codes[sym] = prefix or "0"  # assign "0" if single symbol
        else:
            # Left child gets "0", right child gets "1"
            n_left, l_left, r_left = left
            n_right, l_right, r_right = right
            assign_codes(n_left, l_left, r_left, prefix + "0")
            assign_codes(n_right, l_right, r_right, prefix + "1")
    node0, left0, right0 = tree
    assign_codes(node0, left0, right0, "")
    return [(sym, codes[sym]) for sym, _ in freqs]


__all__ = [
    'my_last', 'my_but_last', 'element_at', 'my_count', 'my_reverse', 'palindrome',
    'flatten_list', 'compress', 'pack', 'encode', 'encode_modified', 'decode',
    'encode_direct', 'dupli', 'repli', 'drop_every', 'split', 'slice_', 'rotate',
    'remove_at', 'insert_at', 'range_list', 'rnd_select', 'lotto_select',
    'rnd_permu', 'combinations', 'group3', 'group', 'lsort', 'lfsort',
    'is_prime', 'gcd', 'coprime', 'totient_phi', 'prime_factors',
    'prime_factors_mult', 'phi_improved', 'nand', 'nor', 'xor', 'impl', 'equ',
    'table', 'assignments', 'table_n', 'gray', 'huffman'
]